File: server/backend/src/models/PowerEvent.js
const { DataTypes } = require('sequelize')
const sequelize = require('../config/database')

const PowerEvent = sequelize.define('PowerEvent', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true
  },
  device_id: {
    type: DataTypes.STRING(50),
    allowNull: false,
    references: {
      model: 'devices',
      key: 'device_id'
    }
  },
  event_type: {
    type: DataTypes.ENUM('power_on', 'power_off', 'battery_low', 'system_error'),
    allowNull: false
  },
  timestamp: {
    type: DataTypes.DATE,
    allowNull: false,
    defaultValue: DataTypes.NOW
  },
  message: {
    type: DataTypes.TEXT,
    allowNull: true
  },
  battery_percentage: {
    type: DataTypes.INTEGER,
    allowNull: true,
    validate: {
      min: 0,
      max: 100
    }
  },
  battery_voltage: {
    type: DataTypes.FLOAT,
    allowNull: true,
    validate: {
      min: 0,
      max: 10
    }
  },
  wifi_signal_strength: {
    type: DataTypes.INTEGER,
    allowNull: true
  },
  free_heap: {
    type: DataTypes.INTEGER,
    allowNull: true
  },
  uptime_ms: {
    type: DataTypes.BIGINT,
    allowNull: true
  },
  metadata: {
    type: DataTypes.JSONB,
    allowNull: true
  }
}, {
  tableName: 'power_events',
  indexes: [
    {
      fields: ['device_id']
    },
    {
      fields: ['timestamp']
    },
    {
      fields: ['event_type']
    },
    {
      fields: ['device_id', 'timestamp']
    },
    {
      fields: ['device_id', 'event_type']
    },
    {
      fields: ['timestamp', 'event_type']
    }
  ],
  hooks: {
    beforeSave: (event) => {
      if (!event.timestamp) {
        event.timestamp = new Date()
      }
    }
  }
})

module.exports = PowerEvent

File: server/backend/models/power_event.go
package models

import "time"

type PowerEvent struct {
	ID        int       `json:"id" db:"id"`
	DeviceID  string    `json:"device_id" db:"device_id"`
	EventType string    `json:"event_type" db:"event_type"`
	Timestamp time.Time `json:"timestamp" db:"timestamp"`
	Data      string    `json:"data,omitempty" db:"data"`
	CreatedAt time.Time `json:"created_at" db:"created_at"`
}

type Device struct {
	ID          string    `json:"id" db:"id"`
	Name        string    `json:"name" db:"name"`
	Description string    `json:"description,omitempty" db:"description"`
	LastSeen    time.Time `json:"last_seen" db:"last_seen"`
	CreatedAt   time.Time `json:"created_at" db:"created_at"`
	UpdatedAt   time.Time `json:"updated_at" db:"updated_at"`
}

type PowerEventRequest struct {
	DeviceID           string    `json:"device_id" binding:"required"`
	Timestamp          time.Time `json:"timestamp"`
	UptimeMs           int64     `json:"uptime_ms"`
	EventType          string    `json:"event_type" binding:"required"`
	Message            string    `json:"message"`
	BatteryPercentage  int       `json:"battery_percentage"`
	BatteryVoltage     float64   `json:"battery_voltage"`
	WiFiSignalStrength int       `json:"wifi_signal_strength"`
	FreeHeap           int64     `json:"free_heap"`
}

type DeviceUpdateRequest struct {
	Name        string `json:"name"`
	Description string `json:"description"`
}
